subprojects/core/src/testFixtures/groovy/org/gradle/internal/classpath/GroovyInterceptorsSubstitution.groovy

```diff
@@ -0,0 +1,135 @@
+/*
+ * Copyright 2023 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.classpath
+
+import org.codehaus.groovy.runtime.callsite.CallSite
+import org.gradle.internal.classpath.intercept.CallInterceptorsSet
+import org.gradle.internal.classpath.intercept.CallSiteDecorator
+
+import javax.annotation.Nullable
+import java.lang.invoke.MethodHandles
+import java.util.concurrent.atomic.AtomicInteger
+
+/**
+ * Provides support for replacing the Groovy interceptors for the current thread.
+ * Make sure that the call sites that should be affected by the change are only reached after the interceptors have been substituted.
+ *
+ * The interceptors are substituted per-thread because other threads may be executing unrelated tests.
+ */
+class GroovyInterceptorsSubstitution {
+    /**
+     * Ensures that the global Groovy call site decorator is replaced with an implementation that maintains call site decorators per thread.
+     * Then sets up the call site decorator for the current thread only.
+     */
+    void setupForCurrentThread() {
+        maybeSetGlobalCallSiteDecorator()
+            .substituteForCurrentThread(
+                new CallInterceptorsSet(GroovyCallInterceptorsProvisionTools.getInterceptorsFromProvider(substitutionProvider).stream())
+            )
+    }
+
+    /**
+     * Cancels the Groovy call site decorator substitution for the current thread.
+     * If the global implementation does not have any other active callsite decorators in the other threads, reverts the global call site
+     * decorator implementation.
+     */
+    static void cleanupForCurrentThread() {
+        substitutionIfPresent()?.cancelSubstitutionForCurrentThread()
+        maybeRevertGlobalCallSiteDecorator()
```

Comment:
I think there's a race between `substituteForCurrentThread` above and `maybeRevertGlobalCallSiteDecorator`. Consider the following order:
0. Thread A calls and completes `setupForCurrentThread`
1. Thread A enters `cleanupForCurrentThread`, compleletes `cancelSubstitutionForCurrentThread`. The `substitutions` counter is 0.
2. Thread B enters `setupForCurrentThread`, completes `maybeSetGlobalCallSiteDecorator`.
3. Thread A completes `maybeRevertGlobalCallSiteDecorator`, restoring the `currentGroovyCallSiteDecorator`
4. Thread B calls `substituteForCurrentThread`, but it has no effect as the `ThreadLocalCallSiteDecoratorSubstitution` is orphaned at this point.

It looks like the lock must be held when modifying the counter for this protocol to work correctly.

Reply:
Ah, indeed. Thanks! Fixed in https://github.com/gradle/gradle/compare/6849a76e891daad3a80cd3e97b529e3093432ea5..f7f2b55d2edb1bbcd0f2d13118ba7fc22401a9d5